### PR TITLE
Add Permissionless Asset Registry Feature

### LINKED ISSUE #892: Make pallet-assets permissionless
Creating assets should be permissionless.
Reference implementation in Statemine.

### PR DESCRIPTION
**Pull request by ferrell-code**
## Description

Can use native token to create an asset for use on manta pay

Makes `AssetRegistryMetadata` a concrete type rather than generic

Creates an additional counter `NextPermissionlessAssetId` so the `AssetIds` of permissionless asset are separate from governance ones.

closes #892 

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [x] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: pallets/asset-manager/src/lib.rs [hunk 1]
```diff
@@ -678,6 +721,77 @@ pub mod pallet {
             });
             Ok(())
         }
+
+        #[pallet::call_index(8)]
+        #[pallet::weight(T::WeightInfo::permissionless_register_asset())]
+        #[transactional]
+        pub fn permissionless_register_asset(
```
Comment (ghzlatarev) [inline pallets/asset-manager/src/lib.rs:728]: We should maybe decouple the mint into its own permissioned_mint function and add a flag for each asset whether  the asset is mintable or not after the initial creation.

File: pallets/asset-manager/src/tests.rs [hunk 2]
```diff
@@ -647,3 +647,166 @@ fn register_lp_asset_should_work() {
...
+        );
+
+        // overflows when decimal is too high
+        assert_noop!(
+            AssetManager::permissionless_register_asset(
+                RuntimeOrigin::signed(ALICE),
+                "dog token".as_bytes().to_vec().try_into().unwrap(),
+                "dog".as_bytes().to_vec().try_into().unwrap(),
+                u8::MAX,
+                u128::MAX,
```
Comment (ghzlatarev) [inline pallets/asset-manager/src/tests.rs]: these 2 can be separate fail cases

Comment (ghzlatarev) [reply pallets/asset-manager/src/tests.rs]: we should also test the limits of the string

Comment (ferrell-code) [reply pallets/asset-manager/src/tests.rs]: a balance of u128::MAX doesn't fail, but agreed it is confusing

File: pallets/asset-manager/src/lib.rs (hunk 1 above)
Comment (ferrell-code) [reply pallets/asset-manager/src/lib.rs:728]: Yeah that could be useful

File: pallets/asset-manager/src/tests.rs (hunk 2 above)
Comment (ferrell-code) [reply pallets/asset-manager/src/tests.rs]: string limit is a little tricky to test as `BoundedVec` with too long of a string won't compile

I suppose trying to decode an encoded call with too long of a string could test

Comment (ghzlatarev) [reply pallets/asset-manager/src/tests.rs]: Oh wait it's BoundedVec, my bad. No need then.

File: pallets/asset-manager/src/lib.rs (hunk 1 above)
Comment (ghzlatarev) [reply pallets/asset-manager/src/lib.rs:728]: Can be separate improvement issue.

**Review by ghzlatarev — DISMISSED** [review]

File: runtime/calamari/src/assets_config.rs [hunk 3]
```diff
@@ -163,18 +163,19 @@ impl AssetIdType for CalamariAssetConfig {
...
     type NativeAssetLocation = NativeAssetLocation;
     type NativeAssetMetadata = NativeAssetMetadata;
-    type StorageMetadata = AssetStorageMetadata;
     type AssetRegistry = CalamariAssetRegistry;
     type FungibleLedger = CalamariConcreteFungibleLedger;
 }
 
 impl pallet_asset_manager::Config for Runtime {
+    type PermissionlessStartId = ConstU128<1_000_000_000>;
+    type TokenNameMaxLen = ConstU32<200>;
```
Comment (Dengjianping) [inline runtime/calamari/src/assets_config.rs]: I think both constants should be aligned with pallet-assets, which has a limit on length of string
```rust
type StringLimit = ConstU32<50>;
```
Under the hood, we manipulate pallet-assets.

File: pallets/asset-manager/src/lib.rs (hunk 1 above)
Comment (Dengjianping) [reply pallets/asset-manager/src/lib.rs:728]: Once user creates an asset, how can it be destroyed?

Comment (ferrell-code) [reply pallets/asset-manager/src/lib.rs:728]: User cannot destroy only a call from `pallet_assets` can do this. So would need to be governance call

File: runtime/calamari/src/assets_config.rs (hunk 3 above)
Comment (ferrell-code) [reply runtime/calamari/src/assets_config.rs]: yeah good call now unified with pallet-assets. It would fail above pallet assets limit

**Comment by Dengjianping** [comment]
Because new extrinsic has been added to this pallet, so you have to add it to tx fee diff test.
```
thread 'diff_tx_fees::diff_tx_fees' panicked at 'assertion failed: `(left == right)`
  left: `9`,
 right: `8`: Please update new extrinsic here.', runtime/calamari/src/diff_tx_fees.rs:1178:9
```

**Comment by ferrell-code** [comment]
> Because new extrinsic has been added to this pallet, so you have to add it to tx fee diff test.
> 
> ```
> thread 'diff_tx_fees::diff_tx_fees' panicked at 'assertion failed: `(left == right)`
>   left: `9`,
>  right: `8`: Please update new extrinsic here.', runtime/calamari/src/diff_tx_fees.rs:1178:9
> ```

Done, also changed the README for the tx fee diff test if you want to take a look at that

**Review by ghzlatarev — DISMISSED** [review]

**Review by ghzlatarev — APPROVED** [review]

**Review by Dengjianping — APPROVED** [review]

**Pull request merged**